OrbitQt/orbittreeview.cpp

```diff
@@ -129,32 +129,58 @@ void OrbitTreeView::OnTimer() {
   }
 }
 
-void OrbitTreeView::Refresh() {
+void OrbitTreeView::Refresh(RefreshMode refresh_mode) {
   if (model_ == nullptr) {
     return;
   }
 
-  QModelIndexList list = selectionModel()->selectedIndexes();
-
   if (model_->GetDataView()->GetType() == DataViewType::kLiveFunctions) {
     model_->layoutAboutToBeChanged();
     model_->layoutChanged();
-    return;
+    // Don't continue the following re-selecting procedure for the live view unless the
+    // refresh is caused by filtering or sorting.
+    if (refresh_mode == RefreshMode::kOther) {
+      return;
+    }
+  } else {
+    reset();
   }
 
-  reset();
-
   // Re-select previous selection
-  int selected = model_->GetSelectedIndex();
-  if (selected >= 0) {
-    QItemSelectionModel* selection = selectionModel();
-    QModelIndex idx = model_->CreateIndex(selected, 0);
-
-    // Don't re-trigger row selection callback when re-selecting.
-    is_internal_refresh_ = true;
-    selection->select(idx, QItemSelectionModel::ClearAndSelect | QItemSelectionModel::Rows);
-    is_internal_refresh_ = false;
+  QItemSelectionModel* selection = selectionModel();
+  QModelIndex index;
+  // Don't re-trigger row selection callback when re-selecting.
+  is_internal_refresh_ = true;
+  if (is_multi_selection_) {
+    std::vector<int> visible_selected_indices = model_->GetDataView()->GetVisibleSelectedIndices();
+    for (int row : visible_selected_indices) {
+      index = model_->CreateIndex(row, 0);
+      selection->select(index, QItemSelectionModel::Select | QItemSelectionModel::Rows);
+    }
+    if (model_->GetDataView()->GetType() == DataViewType::kSampling &&
+        (refresh_mode == RefreshMode::kOnFilter || refresh_mode == RefreshMode::kOnSort)) {
+      // Propagate the re-selection caused by filtering and sorting to the dataview such that the
+      // callstack view will be updated accordingly.
+      model_->GetDataView()->SetIsFilteringOrSorting(true);
+      OnMultiRowsSelected(visible_selected_indices);
+      model_->GetDataView()->SetIsFilteringOrSorting(false);
```

Comment:
Could we set the RefreshMode on the dataview as well, this way SetFilteringOrSorting would become SetRefreshMode() and could be at the beginning and end of the function, reducing code duplication?

Reply:
Change to use `RefreshMode` and `SetRefreshMode` for the `dataview` as well. Thanks for the suggestion!